Allow returning raw bytes from a request, without attempting to decode.

hnefatl opened the pull request:
Some API calls can return binary data, e.g. fetching a user's profile image (e.g. api path like `/api/image/serve/<hash>/512x512`).

Currently this library attempts to decode all responses from `bytes` to `str` using the default `utf-8` encoding, which fails for API calls returning binary data since the first byte of e.g. JPEG data (`0xFF`) can't be decoded to text.

This PR adds a by-default-no-change-to-external-behaviour flag to the `RawClient.request` method to optionally disable decoding for that request, just returning the raw bytes.

hnefatl committed:
Add request param to disable trying to decode all responses to text.

hnefatl committed:
Fix default values

hnefatl commented:
Example usage to fetch a user's profile image to a file:

```
url = "..."
token = "..."
person_id = "person.keith"

with Client(api_url=url, token=token) as client:
    state = client.get_state(entity_id=person_id)
    image_path = state.attributes.get("entity_picture")
    image_path = image_path.removeprefix("/api/")
    with open('out.jpeg', 'wb') as f:
        image_data = client.request(image_path, decode_bytes=False)
        f.write(image_data)
```

GrandMoff100 commented:
Oh! That's actually pretty cool.
I'll review it shortly and we can ship it in the next release in about two weeks.

hnefatl committed:
Fix double-decode

hnefatl committed:
Fix typechecks, after figuring out that they can be run locally using `poetry run mypy homeassistant_api --show-error-codes`

hnefatl committed:
Fix unrelated type error, rather than ignoring.

For some reason, mypy didn't like the typechecking in the ternary operator, but is fine with the statement.

hnefatl commented:
TIL how to use `poetry` properly :) Previous commits weren't done from `poetry shell` so were missing out on the precommit checks.

There was a type error detected by mypy unrelated to my changes (`self._response.method`) which weirdly seemed to be fixed by switching from ternary operator to if statement ¯\\\_(ツ)_/¯

GrandMoff100 commented:
Mypy can be weird like that sometimes lol

GrandMoff100 reviewed (dismissed):
Looks good

hnefatl committed:
Fix typo introduced in 3c9e93a

hnefatl commented:
Sorry, one last tweak - copied the wrong line earlier, thank you type-checking.

GrandMoff100 committed:
Merge branch 'dev' into dev
